modules/scripts/epub3-to-epub3/src/main/java/com/textalk/AppendLegal.java

```diff
@@ -0,0 +1,432 @@
+            <ol>
+                <li>
+                    <a href="DTB36215-000-mtminfo.xhtml#level2_000">[INFORMATION_HEADER]</a>
+                </li>
+            </ol>
+        </li>
+        */
+
+        String fileName = identifier + fileCount + FILE_POSTFIX;
+        Element mainListItem = doc.createElement("li");
+        Element mainAnchor = doc.createElement("a");
+        Element subList = doc.createElement("ol");
+        Element subListItem = doc.createElement("li");
+        Element subAnchor = doc.createElement("a");
+
+        mainAnchor.setAttribute("href", fileName + "#level1_000");
+        mainAnchor.setTextContent(title);
+        subAnchor.setAttribute("href", fileName + "#level2_000");
+        subAnchor.setTextContent(informationHeader);
+
+        mainListItem.appendChild(mainAnchor);
+        mainListItem.appendChild(subList);
+        subList.appendChild(subListItem);
+        subListItem.appendChild(subAnchor);
+        return mainListItem;
+    }
+
+    private int findLevels(Element el) {
+        int levels = 1;
+        NodeList listNodes = el.getChildNodes();
+        for (int i = 0; i < listNodes.getLength(); i++) {
+            if (!"ol".equalsIgnoreCase(listNodes.item(i).getNodeName())) continue;
+            Element olEl = (Element) listNodes.item(i);
+            NodeList listItemNodes = olEl.getChildNodes();
+            for (int j = 0; j < listItemNodes.getLength(); j++) {
+                if (!"li".equalsIgnoreCase(listItemNodes.item(j).getNodeName())) continue;
+                int underLevels = findLevels((Element) listItemNodes.item(j)) + 1;
+                if (levels < underLevels) {
+                    levels = underLevels;
+                }
+            }
+        }
+        return levels;
+    }
+
+    public static void main(String[] args) {
+        AppendLegal al = new AppendLegal();
+        try {
+            al.appendLegalDoc(new File(args[0]), new File(args[1]));
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    ////////// XProc interface //////////
+
+    /**
+     * @author bertfrees
+     */
+    public static class Step extends DefaultStep implements XProcStep {
+
+        @Component(
+            name = "pxi:add-legal-doc",
+            service = { XProcStepProvider.class },
+            property = { "type:String={http://www.daisy.org/ns/pipeline/xproc/internal}add-legal-doc" }
+        )
+        public static class Provider implements XProcStepProvider {
+            @Override
+            public XProcStep newStep(XProcRuntime runtime, XAtomicStep step) {
+                return new Step(runtime, step);
+            }
+        }
+
+        public Step(XProcRuntime runtime, XAtomicStep step) {
+            super(runtime, step);
+        }
+
+        private final static QName INPUT = new QName("input");
+        private final static QName OUTPUT = new QName("input");
```

Comment:
Is this supposed to be input or actually "output" ?

Reply:
Ouch. That's a bug alright. And it might very well be the reason why my XProcSpec test was causing that error! Thanks for catching it.